File: src/main/java/com/VER7U7/UnityPhysics/JUPP/JUPPBridge.java
package com.VER7U7.UnityPhysics.JUPP;

import com.VER7U7.Server.Utils.LittleByteBuffer;

import java.io.*;
import java.net.Socket;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;

public class JUPPBridge {
    private final short port;
    private final AtomicBoolean bridgeRunning = new AtomicBoolean(true);

    private JUPPCommons.BridgeStatus bridgeStatus = JUPPCommons.BridgeStatus.BridgeStarted;

    private static final BlockingQueue<JUPPPacket> outgoingMessageQueue = new LinkedBlockingQueue<>();
    private static final BlockingQueue<JUPPPacket> incomingMessageQueue = new LinkedBlockingQueue<>();

    private Socket socket;
    private Thread updateConnectionThread;
    private Thread readerThread;
    private Thread writerThread;
    private DataInputStream dis;
    private DataOutputStream dos;
    private final Runnable restartCallback;



    public JUPPBridge(short port, Runnable restartCallback) {
        this.port = port;
        this.restartCallback = restartCallback;
    }

    public void startConnection() {
        updateConnectionThread = new Thread(this::updateConnectionLoop, "ConnectionManagerThread");
        updateConnectionThread.start();
    }

    public void stopConnection() {
        bridgeRunning.set(false);
        bridgeStatus = JUPPCommons.BridgeStatus.BridgeStopped;
        try {
            if (socket != null && !socket.isClosed()) {
                socket.close();
            }
        }catch(IOException e) {
            JUPPLog.errprintnln("Error closing socket during stop: " + e.getMessage());
        }
    }

    public void sendPacket(JUPPPacket packet) throws SocketException {
        try {
            outgoingMessageQueue.put(packet);
        }catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            JUPPLog.errprintnln("Send Packet interrupted: " + e.getMessage());
        }
    }

    ///non blocked method
    public JUPPPacket receiveAvalaiblePacket() {
        return incomingMessageQueue.poll();
    }

    public JUPPPacket receivePacket() throws InterruptedException {
        return incomingMessageQueue.take();
    }

    public int receivePoolSize() {
        return incomingMessageQueue.size();
    }

    private void updateConnectionLoop() {
        while (bridgeRunning.get()) {
            try {
                socket = new Socket("localhost", port);
                JUPPLog.println("Connected to Unity at " + socket.getRemoteSocketAddress());
                bridgeStatus = JUPPCommons.BridgeStatus.BridgeConnected;

                dis = new DataInputStream(socket.getInputStream());
                dos = new DataOutputStream(socket.getOutputStream());

                readerThread = new Thread(this::readerThread, "JUPPReaderThread");
                writerThread = new Thread(this::writerThread, "JUPPWriterThread");

                readerThread.start();
                writerThread.start();

                readerThread.join();
                writerThread.join();
            }catch (IOException ioe) {
                JUPPLog.println("Waiting for unity server... (" + ioe.getMessage() + ")");
                try {
                    Thread.sleep(5000);
                } catch(InterruptedException e) {
                    Thread.currentThread().interrupt();
                    JUPPLog.println("Connection manager interrupted while waiting.");
                }
            }catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                JUPPLog.println("Connection manager interrupted while joining threads.");
                bridgeStatus = JUPPCommons.BridgeStatus.BridgeError;
                break;
            }finally {
                closeResources();
                JUPPLog.println("Connection fully closed.");
                bridgeStatus = JUPPCommons.BridgeStatus.BridgeStarted; // Сброс статуса
                outgoingMessageQueue.clear();
                incomingMessageQueue.clear();
                restartCallback.run();
            }
        }
    }

    private void closeResources() {
        try {
            if (dis != null) dis.close();
            if (dos != null) dos.close();
            if (socket != null && !socket.isClosed()) socket.close();
        }catch (IOException e) {
            JUPPLog.errprintnln("Error closing resources: " + e.getMessage());
        } finally {
            dis = null;
            dos = null;
            socket = null;
        }
    }

    private void readerThread() {
        byte[] headerBuffer = new byte[JUPPCommons.JUPP_HEADER.length()];
        byte[] lengthBuffer = new byte[2];

        try {
            while(bridgeRunning.get() && !socket.isClosed()) {
                try {
                    dis.readFully(headerBuffer, 0, JUPPCommons.JUPP_HEADER.length());

                    if (!new String(headerBuffer, StandardCharsets.US_ASCII).equals(JUPPCommons.JUPP_HEADER)) {
                        continue;
                    }

                    dis.readFully(lengthBuffer, 0, 2);
                    ByteBuffer bb = LittleByteBuffer.wrap(lengthBuffer).order(ByteOrder.LITTLE_ENDIAN);
                    short packetLength = bb.getShort();

                    if (packetLength == 0 || packetLength > JUPPCommons.JUPP_PACKET_MAX_SIZE) {
                        continue;
                    }

                    byte[] data = new byte[packetLength];
                    dis.readFully(data, 0, packetLength);

                    handleIncomingPacket(data);
                } catch(EOFException e) {
                    JUPPLog.println("[Client] Disconnected gracefully (EOF).");
                    break;
                } catch (SocketException se) {
                    JUPPLog.println("[Client] Socket Error: " + se.getMessage());
                    se.printStackTrace();
                    JUPPLog.println("[Client] Connection unexpectedly lost.");
                    break;
                } catch (IOException ioe) {
                    JUPPLog.println("[Client] IO Error: " + ioe.getMessage());
                    break;
                }
            }
        } finally {
            JUPPLog.println("JUPPReaderThread is stopped.");
            if (!writerThread.isInterrupted())
                writerThread.interrupt();
        }
    }

    private void writerThread () {
        try {
            while(bridgeRunning.get() && !socket.isClosed()) {
                JUPPPacket packet = outgoingMessageQueue.take();

                byte[] messageToSend = packet.getFormattedData();
                if (messageToSend == null) continue;

                try {
                    byte[] header = JUPPCommons.JUPP_HEADER.getBytes(StandardCharsets.US_ASCII);
                    short messageLength = (short)messageToSend.length;

                    dos.write(header);
                    dos.write(LittleByteBuffer.allocate(2).order(ByteOrder.LITTLE_ENDIAN).putShort(messageLength).array());
                    dos.write(messageToSend);
                    dos.flush();

                    //JUPPLog.println("Packet with id {" + packet.packetID + "} sent successfully.");
                }catch (SocketException se) {
                    JUPPLog.errprintnln("[Client] Socket Error in write loop: " + se.getMessage());
                    JUPPLog.errprintnln("[Client] Connection unexpectedly lost during write.");
                    break;
                } catch(IOException e) {
                    JUPPLog.errprintnln("[Client] IO Error in write loop: " + e.getMessage());
                    break;
                }
            }
        } catch(InterruptedException e) {
            Thread.currentThread().interrupt();
            JUPPLog.println("JUPPWriterThread interrupted.");
        } finally {
            JUPPLog.println("JUPPWriterThread is stopped.");
            if (!readerThread.isInterrupted())
                readerThread.interrupt();
        }
    }

    private void handleIncomingPacket(byte[] rawData) {
        int packetTransferID = LittleByteBuffer.wrap(rawData).order(ByteOrder.LITTLE_ENDIAN).getInt();
        short packetId = LittleByteBuffer.wrap(rawData, 4, 2).order(ByteOrder.LITTLE_ENDIAN).getShort();
        byte[] data = new byte[rawData.length - 6];
        System.arraycopy(rawData, 6, data, 0, data.length);
        JUPPPacket packet = new JUPPPacket(data, packetId, packetTransferID);
        incomingMessageQueue.add(packet);
    }

    public JUPPCommons.BridgeStatus getBridgeStatus() {
        return bridgeStatus;
    }

    public void setBridgeStatus(JUPPCommons.BridgeStatus bridgeStatus) {
        this.bridgeStatus = bridgeStatus;
    }
}

File: JailbreakPhysicsServer/Assets/JUPP/Core/JUPPBridge.cs
using System;
using System.Collections.Concurrent;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Text;
using UnityEngine;

namespace JUPP
{
    public class JUPPBridge
    {

        /*  JUPPBridge - main idea is bridge between Java and Unity by TCP custom protocol
         *  he work only in local network for protection from global space
         *  -- later maybe added global space mode
         */
        private short port;
        private bool _bridgeRunning = true;
        public JUPPCommons.BridgeStatus BridgeStatus { get; private set; }

        private static BlockingCollection<JUPPPacket> _outgoingQueue = new BlockingCollection<JUPPPacket>();
        private static BlockingCollection<JUPPPacket> _incomingQueue = new BlockingCollection<JUPPPacket>();
        
        private TcpListener _listener;
        private TcpClient _client;
        private NetworkStream _stream;
        
        private CancellationTokenSource _cts;
        
        private Thread _updateConnectionThread;
        private Thread _readerThread;
        private Thread _writerThread;
        
        public JUPPBridge(short port)
        {
            this.port = port;
        }

        void OnDisable()
        {
            
        }

        void OnApplicationQuit()
        {
            StopConnection();
        }

        public void StartConnection()
        {
            BridgeStatus = JUPPCommons.BridgeStatus.BridgeStarted;
            
            _updateConnectionThread = new Thread(UpdateConnectionLoop);
            _updateConnectionThread.Start();
        }

        public void StopConnection()
        {
            _bridgeRunning = false;

            try
            {
                CloseResources();
            }
            catch (IOException e)
            {
                JUPPLog.errprintln(e.ToString());
            }
        }

        public void SendPacket(JUPPPacket packet)
        {
            if (packet == null)
                return;
            _outgoingQueue.Add(packet);
        }
        
        
        /// <summary>
        /// Block main thread to get receive packet
        /// </summary>
        public JUPPPacket ReceivePacket()
        {
            return _incomingQueue.Take();
        }
        
        
        /// <summary>
        /// If queue dont have packets return false
        /// </summary>
        public bool TryReceivePacket(out JUPPPacket packet)
        {
            return _incomingQueue.TryTake(out packet);
        }

        public BlockingCollection<JUPPPacket> GetQueuePackets()
        {
            return _incomingQueue;
        }

        public CancellationToken GetCancellationToken()
        {
            return _cts.Token;
        }
        
        public void UpdateConnectionLoop()
        {
            _listener = new TcpListener(IPAddress.Parse("127.0.0.1"), port);
            _listener.Start();
            BridgeStatus = JUPPCommons.BridgeStatus.BridgeSearchConnection;
            
            while (_bridgeRunning)
            {
                try
                {
                    _cts = new CancellationTokenSource();
                    _outgoingQueue = new BlockingCollection<JUPPPacket>();
                    _incomingQueue = new BlockingCollection<JUPPPacket>();
                    
                    _client = _listener.AcceptTcpClient();
                    JUPPLog.println("Connected to JAVA server at " + _client.Client.RemoteEndPoint);
                    BridgeStatus = JUPPCommons.BridgeStatus.BridgeConnected;
                    
                    _stream = _client.GetStream();
                    
                    _readerThread = new Thread(() => ReaderThread(_cts.Token));
                    _writerThread = new Thread(() => WriterThread(_cts.Token));
                    _readerThread.Start();
                    _writerThread.Start();
                    
                    _readerThread.Join();
                    _writerThread.Join();
                }
                catch (SocketException sockEx)
                {
                    if (sockEx.SocketErrorCode == SocketError.Interrupted ||
                        sockEx.SocketErrorCode == SocketError.OperationAborted)
                        JUPPLog.println($"Listener operation cancelled/aborted (likely bridge stopped).");
                    else
                        JUPPLog.println($"Socket error during accept: {sockEx.SocketErrorCode} - {sockEx.Message}");
                }
                catch (ObjectDisposedException)
                {
                    JUPPLog.println($"Listener was disposed.");
                }
                catch (Exception ex)
                {
                    JUPPLog.println($"Unexpected error in main bridge loop: {ex.Message}");
                }
                finally
                {
                    if (_client != null)
                    {
                        JUPPLog.println($"Connection finally closed.");
                        _client.Close();
                    }
                    BridgeStatus = JUPPCommons.BridgeStatus.BridgeSearchConnection;
                }
            }

            CloseResources();
        }

        private void CloseResources()
        {
            try
            {
                
                if (_client != null)
                    _client.Close();
                
                if (_listener != null)
                    _listener.Stop();
            }
            catch (IOException e)
            {
                JUPPLog.errprintln("Error closing resources " + e);
            }
        }

        public void HandleIncomingPacket(byte[] rawData)
        {
            int packetTransferID = BitConverter.ToInt32(rawData, 0);
            short packetId = BitConverter.ToInt16(rawData, 4);
            byte[] packetData = new byte[rawData.Length - 6];
            Buffer.BlockCopy(rawData,  6, packetData, 0, packetData.Length);
            JUPPPacket packet = new JUPPPacket(packetData, packetId, packetTransferID);
            _incomingQueue.Add(packet);
        }

        public void ReaderThread(CancellationToken token)
        {
            byte[] headerBuffer = new byte[JUPPCommons.JUPP_HEADER.Length];
            byte[] lengthBuffer = new byte[2];

            try
            {
                while (_bridgeRunning 
                       && _client.Connected
                       && !token.IsCancellationRequested)
                {
                    try
                    {
                        _stream.ReadExactly(headerBuffer, 0, JUPPCommons.JUPP_HEADER.Length);
                        if (!Encoding.ASCII.GetString(headerBuffer, 0, headerBuffer.Length)
                                .Equals(JUPPCommons.JUPP_HEADER))
                            continue;

                        _stream.ReadExactly(lengthBuffer, 0, 2);
                        short packetLength = BitConverter.ToInt16(lengthBuffer, 0);
                        
                        if (packetLength <= 0 || packetLength > JUPPCommons.JUPP_PACKET_MAX_SIZE)
                        {
                            continue;
                        }

                        byte[] data = new byte[packetLength];
                        _stream.ReadExactly(data, 0, packetLength);
                        
                        HandleIncomingPacket(data);
                    }
                    catch (IOException e)
                    {
                        if (e.InnerException is SocketException se)
                        {
                            if (se.SocketErrorCode == SocketError.ConnectionReset ||
                                se.SocketErrorCode == SocketError.ConnectionAborted)
                                JUPPLog.println(
                                    $"Client {_client.Client.RemoteEndPoint} disconnected unexpectedly (Socket Error: {se.SocketErrorCode}).");
                            else
                                JUPPLog.println(
                                    $"Socket error reading from {_client.Client.RemoteEndPoint}: {se.SocketErrorCode} - {se.Message}");
                        }
                        else if (e.InnerException is EndOfStreamException)
                            JUPPLog.println(
                                $"{_client.Client.RemoteEndPoint} successfully disconnected.");
                        break;
                    }
                    catch (ObjectDisposedException)
                    {
                        JUPPLog.println($"Connection to {_client.Client.RemoteEndPoint} was disposed.");
                        break;
                    }
                    catch (Exception ex)
                    {
                        JUPPLog.println(
                            $"Unexpected error reading from {_client.Client.RemoteEndPoint}: {ex.Message}");
                        break;
                    }
                }
            }
            catch (OperationCanceledException) {}
            finally
            {
                JUPPLog.println("JUPPReaderThread is stopped.");
                if (!_outgoingQueue.IsAddingCompleted)
                    _outgoingQueue.CompleteAdding();
                if (!token.IsCancellationRequested)
                    _cts.Cancel();
            }
        }

        public void WriterThread(CancellationToken token)
        {
            try
            {
                while (
                    _bridgeRunning 
                    && _client.Connected
                    && !token.IsCancellationRequested
                    && _outgoingQueue.TryTake(out JUPPPacket packet, Timeout.Infinite, token))
                {
                    try
                    {
                        byte[] messageToSend = packet.GetFormattedData();
                        if (messageToSend == null) continue;
                        
                        byte[] header = Encoding.ASCII.GetBytes(JUPPCommons.JUPP_HEADER);
                        short messageLength = (short)messageToSend.Length;

                        _stream.Write(header, 0, header.Length);
                        _stream.Write(BitConverter.GetBytes(messageLength));
                        _stream.Write(messageToSend, 0, messageToSend.Length);
                        _stream.Flush();

                        //JUPPLog.println("Packet with id {" + packet.PacketID + "} sent successfully.");
                    }
                    catch (IOException e)
                    {
                        if (e.InnerException is SocketException se)
                        {
                            if (se.SocketErrorCode == SocketError.ConnectionReset ||
                                se.SocketErrorCode == SocketError.ConnectionAborted)
                                JUPPLog.println(
                                    $"Client {_client.Client.RemoteEndPoint} disconnected unexpectedly (Socket Error: {se.SocketErrorCode}).");
                            else
                                JUPPLog.println(
                                    $"Socket error reading from {_client.Client.RemoteEndPoint}: {se.SocketErrorCode} - {se.Message}");
                        }
                        else
                            JUPPLog.println(
                                $"IO Error reading from {_client.Client.RemoteEndPoint}: {e.Message}");

                        break;
                    }
                    catch (ObjectDisposedException)
                    {
                        JUPPLog.println($"Connection to {_client.Client.RemoteEndPoint} was disposed.");
                        break;
                    }
                    catch (InvalidOperationException ioe)
                    {
                        break;
                    }
                    catch (Exception ex)
                    {
                        JUPPLog.println(
                            $"Unexpected error reading from {_client.Client.RemoteEndPoint}: {ex.Message}");
                        break;
                    }
                }
            }
            catch (OperationCanceledException) {}
            finally
            {
                JUPPLog.println("JUPPWriterThread is stopped.");
                if (!_incomingQueue.IsAddingCompleted)
                    _incomingQueue.CompleteAdding();
                
                if (!token.IsCancellationRequested) 
                    _cts.Cancel();
            }
        }
        
    }
}